openpype/hosts/unreal/api/plugin.py

```diff
@@ -1,7 +1,251 @@
 # -*- coding: utf-8 -*-
-from abc import ABC
+import sys
+import six
+from abc import (
+    ABC,
+    ABCMeta,
+)
 
-from openpype.pipeline import LoaderPlugin
+import unreal
+
+from .pipeline import (
+    create_publish_instance,
+    imprint,
+    lsinst,
+    UNREAL_VERSION
+)
+from openpype.lib import (
+    BoolDef,
+    UILabelDef
+)
+from openpype.pipeline import (
+    Creator,
+    LoaderPlugin,
+    CreatorError,
+    CreatedInstance
+)
+
+
+class OpenPypeCreatorError(CreatorError):
+    pass
+
+
+@six.add_metaclass(ABCMeta)
+class UnrealBaseCreator(Creator):
+    """Base class for Unreal creator plugins."""
+    root = "/Game/OpenPype/PublishInstances"
+    suffix = "_INS"
+
+    @staticmethod
+    def cache_subsets(shared_data):
+        """Cache instances for Creators to shared data.
+
+        Create `unreal_cached_subsets` key when needed in shared data and
+        fill it with all collected instances from the scene under its
+        respective creator identifiers.
+
+        If legacy instances are detected in the scene, create
+        `unreal_cached_legacy_subsets` there and fill it with
+        all legacy subsets under family as a key.
+
+        Args:
+            Dict[str, Any]: Shared data.
+
+        Return:
+            Dict[str, Any]: Shared data dictionary.
+
+        """
+        if shared_data.get("unreal_cached_subsets") is None:
+            shared_data["unreal_cached_subsets"] = {}
+            if shared_data.get("unreal_cached_legacy_subsets") is None:
+                shared_data["unreal_cached_legacy_subsets"] = {}
+            cached_instances = lsinst()
+            for i in cached_instances:
+                if not i.get("creator_identifier"):
+                    # we have legacy instance
+                    family = i.get("family")
+                    if (family not in
+                            shared_data["unreal_cached_legacy_subsets"]):
+                        shared_data[
+                            "unreal_cached_legacy_subsets"][family] = [i]
+                    else:
+                        shared_data[
+                            "unreal_cached_legacy_subsets"][family].append(i)
+                    continue
+
+                creator_id = i.get("creator_identifier")
+                if creator_id not in shared_data["unreal_cached_subsets"]:
+                    shared_data["unreal_cached_subsets"][creator_id] = [i]
+                else:
+                    shared_data["unreal_cached_subsets"][creator_id].append(i)
```

Comment:
This could be simplified to
```suggestion
            unreal_cached_subsets = collections.defaultdict(list)
            unreal_cached_legacy_subsets = collections.defaultdict(list)
            for instance in lsinst():
                creator_id = instance.get("creator_identifier")
                if creator_id:
                    unreal_cached_subsets[creator_id].append(instance)
                else:
                    family = instance.get("family")
                    unreal_cached_legacy_subsets[family].append(instance)

            shared_data["unreal_cached_subsets"] = unreal_cached_subsets
            shared_data["unreal_cached_legacy_subsets"] = (
                unreal_cached_legacy_subsets
            )
```